// @gumhq/ui-components/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GumUIProvider = exports.GumUI = exports.Feed = exports.FollowButton = exports.Profile = exports.Post = void 0;
const GumUI = __importStar(require("@nextui-org/react"));
exports.GumUI = GumUI;
const react_1 = require("@nextui-org/react");
Object.defineProperty(exports, "GumUIProvider", { enumerable: true, get: function () { return react_1.NextUIProvider; } });
var post_1 = require("./components/post");
Object.defineProperty(exports, "Post", { enumerable: true, get: function () { return post_1.Post; } });
var profile_1 = require("./components/profile");
Object.defineProperty(exports, "Profile", { enumerable: true, get: function () { return profile_1.Profile; } });
var buttons_1 = require("./components/buttons");
Object.defineProperty(exports, "FollowButton", { enumerable: true, get: function () { return buttons_1.FollowButton; } });
var feed_1 = require("./components/feed");
Object.defineProperty(exports, "Feed", { enumerable: true, get: function () { return feed_1.Feed; } });
